google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkUpdateService.java

```diff
@@ -0,0 +1,140 @@
+      return;
+    }
+
+    if (updateTimer == null) {
+      updateTimer = new Timer(SDK_UPDATER_THREAD_NAME);
+    }
+    // cancel tasks from previous activation.
+    if (sdkUpdateTask != null) {
+      sdkUpdateTask.cancel();
+    }
+
+    sdkUpdateTask =
+        new TimerTask() {
+          @Override
+          public void run() {
+            doUpdate();
+          }
+        };
+    long lastTimeOfUpdate =
+        CloudSdkServiceUserSettings.getInstance().getLastAutomaticUpdateTimestamp();
+    long delayBeforeCheckMillis = SDK_UPDATE_INTERVAL_MS;
+    long now = getClock().millis();
+    if (lastTimeOfUpdate != 0 && lastTimeOfUpdate < now) {
+      delayBeforeCheckMillis =
+          Math.min(
+              SDK_UPDATE_INTERVAL_MS,
+              Math.max(0, SDK_UPDATE_INTERVAL_MS - (getClock().millis() - lastTimeOfUpdate)));
+    }
+
+    schedule(sdkUpdateTask, delayBeforeCheckMillis, SDK_UPDATE_INTERVAL_MS);
+  }
+
+  /** Called when managed SDK update operation (update or install) successfully completes. */
+  void notifySdkUpdate() {
+    CloudSdkServiceUserSettings.getInstance().setLastAutomaticUpdateTimestamp(getClock().millis());
+  }
+
+  @VisibleForTesting
+  Clock getClock() {
+    return Clock.systemDefaultZone();
+  }
+
+  @VisibleForTesting
+  void schedule(TimerTask timerTask, long delay, long period) {
+    updateTimer.scheduleAtFixedRate(timerTask, delay, period);
+  }
+
+  @VisibleForTesting
+  javax.swing.Timer createUiTimer(int delayMillis) {
+    return new javax.swing.Timer(delayMillis, null);
+  }
+
+  private void doUpdate() {
+    CloudSdkService cloudSdkService = CloudSdkService.getInstance();
+    if (cloudSdkService instanceof ManagedCloudSdkService) {
+      // do not show notifications and start update process if SDK is still up-to-date.
+      boolean updateFeatureEnabled =
+          ServiceManager.getService(PluginInfoService.class)
+              .shouldEnable(GctFeature.MANAGED_SDK_UPDATE);
+
+      if (updateFeatureEnabled && !((ManagedCloudSdkService) cloudSdkService).isUpToDate()) {
+        // schedule UI timer to let a user decide about proceeding with the update.
+        javax.swing.Timer uiTimer = createUiTimer(UPDATE_NOTIFICATION_DELAY_MS);
+        ActionListener cancelListener = (e) -> uiTimer.stop();
+        ActionListener disableUpdateListener =
+            (e) -> {
+              uiTimer.stop();
+              sdkUpdateTask.cancel();
+              CloudSdkServiceUserSettings.getInstance().setEnableAutomaticUpdates(false);
+            };
+        Notification updateNotification =
+            ManagedCloudSdkServiceUiPresenter.getInstance()
+                .notifyManagedSdkUpdate(cancelListener, disableUpdateListener);
+
+        uiTimer.addActionListener(
+            e -> {
+              updateNotification.expire();
+              // grab a current SDK service, might have changed while waiting for notification.
```

Comment:
interesting. so this is to guard against the user quickly changing the SDK to be custom (instead of managed) in the interval before the notification expired? In that case, it will just skip the update right?

Reply:
Correct, that's the case where we already scheduled to be updated and then a user changes the SDK to custom, then we skip and we it stays custom, next scheduled run will cancel the update task completely.